AX: [AX Thread Text APIs] CurrentLine is sometimes missing the first character of a line

hoffmanjoshua opened the pull request:
#### e67fcd568d5f0ea91b202f58514c73549dbe5ce2
<pre>
AX: [AX Thread Text APIs] CurrentLine is sometimes missing the first character of a line
<a href="https://bugs.webkit.org/show_bug.cgi?id=287438">https://bugs.webkit.org/show_bug.cgi?id=287438</a>
<a href="https://rdar.apple.com/143915387">rdar://143915387</a>

Reviewed by Tyler Wilcock.

Due to some bugs in `findLine`, the current line range would sometimes miss the first character
in a line. This is because `atLineBoundaryForDirection` didn&apos;t handle text affinity, so if we
have the following text, with a soft line break after hello:

Hello
world

A text marker after the &apos;w&apos; in world would be considered at a line break with our old implementation,
which would just check the previous text marker&apos;s line (which could be on the previous line). This PR
re-writes `atLineBoundaryForDirection` to use line offsets instead, which should be more robust
against these affinity cases.

This PR also includes some text marker enhancements:
- Iterating in `findLine` now follows the direction of the search, which is more accurate and
aligned with our other `find` methods.
- A new debug method `printTrees` was added to the WKTR accessibility controller to more easily
log the AX trees to standard error when debugging.
    - This change includes added a new optional paramater to treeData for streaming options.
- More TextMarkerOrigin fields were added, with new parameters where necessary to pipe them
through.
- An update to `focusedLocalFrame`, co-authored by Alex Christensen, to fix a memory-safety
issue with accessing m_page.

* LayoutTests/accessibility/ax-thread-text-apis/line-range-at-soft-breaks-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/line-range-at-soft-breaks.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForCharacterOffset):
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
(WebCore::AXObjectCache::treeData):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::atLineBoundaryForDirection const):
(WebCore::AXTextMarker::findLine const):
(WebCore::AXTextMarker::lineRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::originToString):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::textMarkerForCharacterOffset):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityPerformAction:]):
(-[WebAccessibilityObjectWrapper _accessibilityPrintTrees]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase focusedLocalFrame]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.h:
(WTR::AccessibilityController::printTrees):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityController.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
(WTR::AccessibilityController::printTrees):

Canonical link: <a href="https://commits.webkit.org/290212@main">https://commits.webkit.org/290212@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5a9da33ff398d1d45f4cb3ecb12c392ee4d83e4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8779 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44087 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94238 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40016 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91306 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17078 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68766 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26440 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92257 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7029 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81004 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49127 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35384 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39123 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77137 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36372 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96070 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16435 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77643 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76791 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76938 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18986 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21352 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19960 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9555 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16449 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21760 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16190 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19641 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17971 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -806,19 +838,20 @@ AXTextMarker AXTextMarker::findLine(AXDirection direction, AXTextUnitBoundary bo
         unsigned cumulativeOffset = runIndex ? currentRuns->runLengthSumTo(runIndex - 1) : 0;
-        for (size_t i = runIndex; i < currentRuns->size(); i++) {
+        // We should search in the right direction for a change in the line index.
+        for (int i = runIndex; direction == AXDirection::Next ? i < (int)currentRuns->size() : i >= 0; direction == AXDirection::Next ? i++ : i--) {
```
If we used `unsigned i` instead of `int i`, could we avoid the `(int)` cast?

hoffmanjoshua replied:
If we used unsigned, when the direction is `Previous`, we would never terminate (since `i` would wrap around). We could make this cast a `Checked<int>` though?

twilco replied:
Ahh right. Another option would be to add this at the bottom of the loop body:

```cpp
if (direction == AXDirection::Previous && !i) {
    // We want to execute the loop body when i == 0, but break now to avoid underflow.
    break;
}
```

What do you think is best?

hoffmanjoshua replied:
I don't mind that break condition. I'll make `runIndex` a `size_t` while I'm at it, to match what we use for its type everywhere else

twilco commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -335,6 +347,8 @@ class AXTextMarker {
     AXTextRunLineID lineID() const;
     // Are we at the start or end of a line?
     bool atLineBoundaryForDirection(AXDirection) const;
+    // "Fast path" to calcuate line boundary when a callsite already has the runs and runIndex available.
```
Are the quotes around "Fast path" necessary? The quotes imply dubiousness around the term (to me), which seems unnecessary, but doesn't matter too much.

hoffmanjoshua replied:
Definitely not necessary, I'll remove those 😅

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2448,6 +2448,25 @@ - (void)accessibilityPerformAction:(NSString*)action
         [self accessibilityScrollToVisible];
     else if ([action isEqualToString:@"AXDismissAction"])
         backingObject->performDismissActionIgnoringResult();
+    else if ([action isEqualToString:@"AXLogTrees"])
```
Should we only support this action for test clients (e.g. in the context of a layout test)? Or do you use it when debugging in Safari / MiniBrowser too?

hoffmanjoshua replied:
We can limit this to test clients—that's the only place I've been using this

hoffmanjoshua committed:
AX: [AX Thread Text APIs] CurrentLine is sometimes missing the first character of a line
https://bugs.webkit.org/show_bug.cgi?id=287438
rdar://143915387

Reviewed by Tyler Wilcock.

Due to some bugs in `findLine`, the current line range would sometimes miss the first character
in a line. This is because `atLineBoundaryForDirection` didn't handle text affinity, so if we
have the following text, with a soft line break after hello:

Hello
world

A text marker after the 'w' in world would be considered at a line break with our old implementation,
which would just check the previous text marker's line (which could be on the previous line). This PR
re-writes `atLineBoundaryForDirection` to use line offsets instead, which should be more robust
against these affinity cases.

This PR also includes some text marker enhancements:
- Iterating in `findLine` now follows the direction of the search, which is more accurate and
aligned with our other `find` methods.
- A new debug method `printTrees` was added to the WKTR accessibility controller to more easily
log the AX trees to standard error when debugging.
    - This change includes added a new optional paramater to treeData for streaming options.
- More TextMarkerOrigin fields were added, with new parameters where necessary to pipe them
through.
- An update to `focusedLocalFrame`, co-authored by Alex Christensen, to fix a memory-safety
issue with accessing m_page.

* LayoutTests/accessibility/ax-thread-text-apis/line-range-at-soft-breaks-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/line-range-at-soft-breaks.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForCharacterOffset):
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
(WebCore::AXObjectCache::treeData):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::atLineBoundaryForDirection const):
(WebCore::AXTextMarker::findLine const):
(WebCore::AXTextMarker::lineRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::originToString):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::textMarkerForCharacterOffset):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityPerformAction:]):
(-[WebAccessibilityObjectWrapper _accessibilityPrintTrees]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase focusedLocalFrame]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.h:
(WTR::AccessibilityController::printTrees):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityController.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
(WTR::AccessibilityController::printTrees):

Canonical link: https://commits.webkit.org/290212@main
